### ISSUE TITLE
Add support for wrapping a ByteBuffer as ByteString without copy.

### ISSUE THREAD

**Issue by hepin1989** [akka/akka#25160]
My friend is preparing a hackathon with Akka, and found Akka is doing copy every time it emits a `Received(ByteString(buffer))` message。

Akka should support wrapping a ByteBuffer(may from the pool) as ByteString without any copy and can do a manually release once it's not needed, otherwise there may be a leak.

**Comment by hepin1989** [comment]
https://github.com/akka/akka/issues/17281
https://github.com/akka/akka/issues/24556

**Comment by hepin1989** [comment]
Another thing emphasized here is: Could we redesign the zero-copy of ByteString as a whole?

**Comment by johanandren** [comment]
While I can see how this would be desirable from a performance perspective, changing the immutable `ByteString` into something with a lifecycle that needs to be released at some point doesn't seem possible in my opinion. This would mean that every piece of user code that uses it would turn into a potential memory leak (in addition to the immutability guarantee being so much weaker).

For some places we might be able to do that internally, in Artery for example, since there the bytes never leave our control until they are turned into something else through serialization.

**Comment by hepin1989** [comment]
Yes, that's why Netty make use of an `io.netty.util.ResourceLeakDetector` and need the user to turn it on in development.

**Comment by WayneWang12** [comment]
It seems I can wrap the `java.nio.ByteBuffer` using netty `ByteBuf`. I've tried it on my own project. I'll see if it works. If it does, then I think akka can provide one more type of `Tcp` or `Udp` which directly provide `ByteBuf` or some other  zero-copy buffer for experienced users to let them build up more efficient infrastructures. I'll report my finding in a few weeks.

**Comment by hepin1989** [comment]
I  think at first we should abstract away with ByteBuffer?

![image](https://user-images.githubusercontent.com/501740/40713683-1dc93cd8-6433-11e8-822e-ee875e2c1b13.png)

And make use of `ByteString` here?

**Comment by ktoso** [comment]
I'm not sure what you are trying to achieve here...
We all know and like Netty, however ByteString is NOT ByteBuf and shall not become such structure -- it must be freely shareable without caring if it's backed by something that has to be released later on or not.

Note also that ByteString is in akka-actor which must not depend on Netty, as akka-actor may not depend on more than an absolute minimum -- which it currently does.

Please before spending much time on this, clarify goals and only then get hacking, I would not want you to spend much time and effort on a thing which on principle could not be included in Akka later on then.

**Comment by WayneWang12** [comment]
@ktoso Just attending a hackthon held by Alibaba...I want to use akka to make a good rank. But the hackthon requires ultimate performance in network, which in akka seems cannot beat netty...

Yeah, I've found it is difficult. Memory leak is everywhere. Also have to do `Framing` all by myself. But if I finished it, I'll just use the hacked one in the game. Certainly not in akka. It's really hard to control.

I just can't take the fact that akka is not as good as netty in network performance... For such a long time, I thought akka is always the best in every aspects...I'll try my best to beat netty in this hackthon.

**Comment by hepin1989** [comment]
@ktoso This need be well thought and designed, I opened up this issue as a holder, so could latter act on it. The current Akka-io is very good, just could not achieve the ultrafast performance for some dedicated special use case. In the inner hackton, we beat service mesh agent which wroten in golang ,c++ ,c and etc, So open this kind of issue is want to make Akka better and better. And this may come up with some PRs, the Akka team have some more higher priority works to do.

So, this is kind of *exploring* thing:), Happy Hakking.
